RL4J Added listener pattern to SyncLearning

aboulang2002 opened the pull request:
Signed-off-by: Alexandre Boulanger <aboulang2002@yahoo.com>

## What changes were proposed in this pull request?

The goal of this PR is to remove IDataManager from SyncLearning. I added a listener pattern to SyncLearning, and a listener class (DataManagerSyncTrainingListener) that will make the calls to DataManager that SyncLearning used to do.

I removed the try/catch from SyncLearning.train(). Listeners should be well-behaved and not throw. If a listener need to stop the training (because of an IOException for example), it can with SyncTrainingEvent.setCanContinue(false).

Since SyncLearning doesn't use IDataManager anymore, I moved it to ASyncLearning (where a similar refac will be done in an upcoming PR).

I did not add javadoc to IDataManager and IAsyncGlobal.
In the case of IDataManager/DataManager, it may be worthwhile to refac it. For example, isSaveData() will soon not be necessary anymore. Also, it seems strange to me that a DataManager would return the path where to save the videos which are then saved by a HistoryProcessor. I will write the javadoc when I change this class.
In the case of IAsyncGlobal, I don't understand a few things and I'd like to take the time to read and understand "Hogwild!" before writing the javadoc.

Summary of changes:
 - Moved up IDataManager from Learning to ASyncLearning and removed it from SyncLearning
 - Some code in QLearningDiscrete.postEpoch() has been moved to DataManagerSyncTrainingListener.onEpochEnd()
 - Added listener pattern to SyncLearning along with a listener, DataManagerSyncTrainingListener.
 - Adjusted SyncLearningTest for the new listener
 - Added refac tests to QLearningDiscrete to make sure it remains functionally the same as before.

## How was this patch tested?

Unit tests

## Quick checklist

The following checklist helps ensure your PR is complete:

- [X] Eclipse Contributor Agreement signed, and signed commits - see [IP Requirements](https://deeplearning4j.org/eclipse-contributors) page for details
- [X] Reviewed the [Contributing Guidelines](https://github.com/eclipse/deeplearning4j/blob/master/CONTRIBUTING.md) and followed the steps within.
- [X] Created tests for any significant new code additions.
- [X] Relevant tests for your changes are passing.

saudet commented on rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/sync/listener/SyncTrainingListener.java:
```diff
@@ -0,0 +1,31 @@
+ */
+public interface SyncTrainingListener {
+    /**
+     * Called once when the training starts. The training can be aborted by calling event.setCanContinue(false)
```
@AlexDBlack Are you OK with using a listener for that?

AlexDBlack replied:
Yes, we're actually looking at a similar design in SameDiff listeners.
Though we're going with something like onEpochEnd returning an enumeration - `ListenerResponse.STOP` or `ListenerResponse.CONTINUE` sort of thing.

saudet commented on rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/sync/qlearning/discrete/QLearningDiscrete.java:
```diff
@@ -68,24 +66,29 @@
-
 
+    /**
+     * Obsolete. Use QLearningDiscrete(MDP, IDQN, QLConfiguration, int) with addListener() instead.
```
Let's mark those with `@Deprecated`.

saudet commented on rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/sync/SyncLearning.java:
```diff
@@ -35,50 +41,93 @@
+        log.info("training starting.");
 
+        boolean canContinue = notifyTrainingStarted();
+        if(canContinue) {
```
The formatting looks off in this file. Let's be as consistent as possible. :)

AlexDBlack commented on rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/util/DataManagerSyncTrainingListener.java:
```diff
@@ -0,0 +1,105 @@
+            dataManager.writeInfo(event.getLearning());
+        } catch (Exception e) {
+            log.error("Training failed.", e);
+            e.printStackTrace();
```
Don't also print stack trace, log.error will already print it.
Also other instances below.

aboulang2002 commented:
Summary of changes:
- Added/changed javadocs
- Make listeners return a response instead of changing a canContinue property (also changed SyncLearning and unit tests)
- removed all e.printStackTrace() in DataManagerSyncTrainingListener

saudet commented:
Looks good to me! Tested it out a bit, and it doesn't appear to break anything.
